### PR TITLE
layers: Fix CMD_BUFFER_STATE NotifyInvalidate

### LINKED ISSUE #5011: DynamicRenderingNullDepthStencilExecuteCommands | Deadlock
**Describe the Issue**
```
[ RUN      ] VkLayerTest.DynamicRenderingNullDepthStencilExecuteCommands
==================
WARNING: ThreadSanitizer: lock-order-inversion (potential deadlock) (pid=15335)
  Cycle in lock order graph: M0 (0x7b740001df80) => M1 (0x7b740001e980) => M0
```

Run `DynamicRenderingNullDepthStencilExecuteCommands` with a thread sanitized build.

<details>

<summary>Full ASAN Log</summary>

```
[ RUN      ] VkLayerTest.DynamicRenderingNullDepthStencilExecuteCommands
==================
WARNING: ThreadSanitizer: lock-order-inversion (potential deadlock) (pid=16833)
  Cycle in lock order graph: M0 (0x7b740001df80) => M1 (0x7b740001e980) => M0

  Mutex M1 acquired here while holding mutex M0 in main thread:
    #0 pthread_rwlock_rdlock <null> (vk_layer_validation_tests+0x31688a) (BuildId: 655a6717424f7bc5707081155d2c11ce71651f2f)
    #1 std::__glibcxx_rwlock_rdlock(pthread_rwlock_t*) /usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/shared_mutex:78:3 (libVkLayer_khronos_validation.so+0x1898f13) (BuildId: 036adb65bc27970aa02946267092770927a02df7)
    #2 std::__shared_mutex_pthread::lock_shared() /usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/shared_mutex:229:10 (libVkLayer_khronos_validation.so+0x191295d) (BuildId: 036adb65bc27970aa02946267092770927a02df7)
    #3 std::shared_mutex::lock_shared() /usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/shared_mutex:426:34 (libVkLayer_khronos_validation.so+0x1912905) (BuildId: 036adb65bc27970aa02946267092770927a02df7)
    #4 std::shared_lock<std::shared_mutex>::shared_lock(std::shared_mutex&) /usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/shared_mutex:727:13 (libVkLayer_khronos_validation.so+0x19128ab) (BuildId: 036adb65bc27970aa02946267092770927a02df7)
    #5 CMD_BUFFER_STATE::ReadLock() const /home/juan/projects/vvl/layers/cmd_buffer_state.h:340:45 (libVkLayer_khronos_validation.so+0x1f5c77c) (BuildId: 036adb65bc27970aa02946267092770927a02df7)
    #6 LockedSharedPtr<CMD_BUFFER_STATE const, std::shared_lock<std::shared_mutex> > ValidationStateTracker::GetRead<CMD_BUFFER_STATE, state_object::Traits<CMD_BUFFER_STATE>, LockedSharedPtr<CMD_BUFFER_STATE const, std::shared_lock<std::shared_mutex> > >(state_object::Traits<CMD_BUFFER_STATE>::HandleType) const /home/juan/projects/vvl/layers/state_tracker.h:417:31 (libVkLayer_khronos_validation.so+0x1f0d93c) (BuildId: 036adb65bc27970aa02946267092770927a02df7)
    #7 CoreChecks::PreCallValidateCmdExecuteCommands(VkCommandBuffer_T*, unsigned int, VkCommandBuffer_T* const*) const /home/juan/projects/vvl/layers/core_validation.cpp:16394:29 (libVkLayer_khronos_validation.so+0x1ea22a1) (BuildId: 036adb65bc27970aa02946267092770927a02df7)
    #8 vulkan_layer_chassis::CmdExecuteCommands(VkCommandBuffer_T*, unsigned int, VkCommandBuffer_T* const*) /home/juan/projects/vvl/layers/generated/chassis.cpp:3748:67 (libVkLayer_khronos_validation.so+0x18084d8) (BuildId: 036adb65bc27970aa02946267092770927a02df7)
    #9 VkLayerTest_DynamicRenderingNullDepthStencilExecuteCommands_Test::TestBody() /home/juan/projects/vvl/tests/vklayertests_dynamic_rendering.cpp:5610:5 (vk_layer_validation_tests+0xb9a4c8) (BuildId: 655a6717424f7bc5707081155d2c11ce71651f2f)
    #10 void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) /home/juan/projects/vvl/external/googletest/googletest/src/gtest.cc:2599:10 (vk_layer_validation_tests+0x12b309a) (BuildId: 655a6717424f7bc5707081155d2c11ce71651f2f)
    #11 main /home/juan/projects/vvl/tests/layer_validation_tests.cpp:3758:14 (vk_layer_validation_tests+0x3b6c9c) (BuildId: 655a6717424f7bc5707081155d2c11ce71651f2f)

    Hint: use TSAN_OPTIONS=second_deadlock_stack=1 to get more informative warning message

  Mutex M0 acquired here while holding mutex M1 in main thread:
    #0 pthread_rwlock_wrlock <null> (vk_layer_validation_tests+0x316c5a) (BuildId: 655a6717424f7bc5707081155d2c11ce71651f2f)
    #1 std::__glibcxx_rwlock_wrlock(pthread_rwlock_t*) /usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/shared_mutex:80:3 (libVkLayer_khronos_validation.so+0x1898f73) (BuildId: 036adb65bc27970aa02946267092770927a02df7)
    #2 std::__shared_mutex_pthread::lock() /usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/shared_mutex:193:19 (libVkLayer_khronos_validation.so+0x1912b95) (BuildId: 036adb65bc27970aa02946267092770927a02df7)
    #3 std::shared_mutex::lock() /usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/shared_mutex:420:27 (libVkLayer_khronos_validation.so+0x1912b45) (BuildId: 036adb65bc27970aa02946267092770927a02df7)
    #4 std::unique_lock<std::shared_mutex>::lock() /usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/bits/unique_lock.h:139:17 (libVkLayer_khronos_validation.so+0x1912ad5) (BuildId: 036adb65bc27970aa02946267092770927a02df7)
    #5 std::unique_lock<std::shared_mutex>::unique_lock(std::shared_mutex&) /usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/bits/unique_lock.h:69:2 (libVkLayer_khronos_validation.so+0x1912a17) (BuildId: 036adb65bc27970aa02946267092770927a02df7)
    #6 CMD_BUFFER_STATE::WriteLock() /home/juan/projects/vvl/layers/cmd_buffer_state.h:341:41 (libVkLayer_khronos_validation.so+0x1ce6afc) (BuildId: 036adb65bc27970aa02946267092770927a02df7)
    #7 CMD_BUFFER_STATE::NotifyInvalidate(small_vector<std::shared_ptr<BASE_NODE>, 4ul, unsigned int> const&, bool) /home/juan/projects/vvl/layers/cmd_buffer_state.cpp:289:22 (libVkLayer_khronos_validation.so+0x1f9be5e) (BuildId: 036adb65bc27970aa02946267092770927a02df7)
    #8 BASE_NODE::NotifyInvalidate(small_vector<std::shared_ptr<BASE_NODE>, 4ul, unsigned int> const&, bool) /home/juan/projects/vvl/layers/base_node.cpp:108:19 (libVkLayer_khronos_validation.so+0x1f8e30c) (BuildId: 036adb65bc27970aa02946267092770927a02df7)
    #9 BASE_NODE::Invalidate(bool) /home/juan/projects/vvl/layers/base_node.cpp:94:16 (libVkLayer_khronos_validation.so+0x1f8d903) (BuildId: 036adb65bc27970aa02946267092770927a02df7)
    #10 CMD_BUFFER_STATE::Reset() /home/juan/projects/vvl/layers/cmd_buffer_state.cpp:183:5 (libVkLayer_khronos_validation.so+0x1f9b165) (BuildId: 036adb65bc27970aa02946267092770927a02df7)
    #11 CMD_BUFFER_STATE::Destroy() /home/juan/projects/vvl/layers/cmd_buffer_state.cpp:282:9 (libVkLayer_khronos_validation.so+0x1f9bd20) (BuildId: 036adb65bc27970aa02946267092770927a02df7)
    #12 void ValidationStateTracker::Destroy<CMD_BUFFER_STATE, state_object::Traits<CMD_BUFFER_STATE> >(state_object::Traits<CMD_BUFFER_STATE>::HandleType) /home/juan/projects/vvl/layers/state_tracker.h:341:27 (libVkLayer_khronos_validation.so+0x1fabf92) (BuildId: 036adb65bc27970aa02946267092770927a02df7)
    #13 COMMAND_POOL_STATE::Free(unsigned int, VkCommandBuffer_T* const*) /home/juan/projects/vvl/layers/cmd_buffer_state.cpp:53:23 (libVkLayer_khronos_validation.so+0x1f99a15) (BuildId: 036adb65bc27970aa02946267092770927a02df7)
    #14 ValidationStateTracker::PreCallRecordFreeCommandBuffers(VkDevice_T*, VkCommandPool_T*, unsigned int, VkCommandBuffer_T* const*) /home/juan/projects/vvl/layers/state_tracker.cpp:2241:15 (libVkLayer_khronos_validation.so+0x1c6f9c0) (BuildId: 036adb65bc27970aa02946267092770927a02df7)
    #15 vulkan_layer_chassis::FreeCommandBuffers(VkDevice_T*, VkCommandPool_T*, unsigned int, VkCommandBuffer_T* const*) /home/juan/projects/vvl/layers/generated/chassis.cpp:2645:20 (libVkLayer_khronos_validation.so+0x17f8732) (BuildId: 036adb65bc27970aa02946267092770927a02df7)
    #16 vk_testing::CommandBuffer::destroy() /home/juan/projects/vvl/tests/vktestbinding.cpp:1063:5 (vk_layer_validation_tests+0xfb796e) (BuildId: 655a6717424f7bc5707081155d2c11ce71651f2f)
    #17 vk_testing::CommandBuffer::~CommandBuffer() /home/juan/projects/vvl/tests/vktestbinding.cpp:1066:44 (vk_layer_validation_tests+0xfb79c5) (BuildId: 655a6717424f7bc5707081155d2c11ce71651f2f)
    #18 VkCommandBufferObj::~VkCommandBufferObj() /home/juan/projects/vvl/tests/vkrenderframework.h:363:7 (vk_layer_validation_tests+0x3be365) (BuildId: 655a6717424f7bc5707081155d2c11ce71651f2f)
    #19 VkLayerTest_DynamicRenderingNullDepthStencilExecuteCommands_Test::TestBody() /home/juan/projects/vvl/tests/vklayertests_dynamic_rendering.cpp:5651:1 (vk_layer_validation_tests+0xb9aa9d) (BuildId: 655a6717424f7bc5707081155d2c11ce71651f2f)
    #20 void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) /home/juan/projects/vvl/external/googletest/googletest/src/gtest.cc:2599:10 (vk_layer_validation_tests+0x12b309a) (BuildId: 655a6717424f7bc5707081155d2c11ce71651f2f)
    #21 main /home/juan/projects/vvl/tests/layer_validation_tests.cpp:3758:14 (vk_layer_validation_tests+0x3b6c9c) (BuildId: 655a6717424f7bc5707081155d2c11ce71651f2f)

SUMMARY: ThreadSanitizer: lock-order-inversion (potential deadlock) (/home/juan/projects/vvl/build/tests/vk_layer_validation_tests+0x31688a) (BuildId: 655a6717424f7bc5707081155d2c11ce71651f2f) in __interceptor_pthread_rwlock_rdlock
==================
[       OK ] VkLayerTest.DynamicRenderingNullDepthStencilExecuteCommands (6522 ms)
```

</details>

### PR DESCRIPTION
**Pull request by juan-lunarg**
closes #5011

### CODE REVIEW

**Comment by juan-lunarg** [comment]
1.) CMD_BUFFER_STATE::Destroy() | starts a WriteLock, now CMD_BUFFER_STATE's lock is in use
2.) CMD_BUFFER_STATE::Reset()
3.) BASE_NODE::Invalidate
4.) BASE_NODE::NotifyInvalidate
5.) CMD_BUFFER_STATE::NotifyInvalidate | tries to start a WriteLock, with the already in use lock

I believe the solution is to remove the lock from `NotififyInvalidate`. AFAIK there is no issue with this approach. Let me know what you think @jeremy-lunarg

File: layers/cmd_buffer_state.cpp [hunk 1]
```diff
@@ -285,47 +285,44 @@ void CMD_BUFFER_STATE::Destroy() {
...
+    for (auto &obj : invalid_nodes) {
+        log_list.add(obj->Handle());
+    }
+
+    bool found_invalid = false;
+    for (auto &obj : invalid_nodes) {
+        // Only record a broken binding if one of the nodes in the invalid chain is still
+        // being tracked by the command buffer. This is to try to avoid race conditions
+        // caused by separate CMD_BUFFER_STATE and BASE_NODE::parent_nodes locking.
+        if (object_bindings.erase(obj)) {
```
Comment (jeremyg-lunarg) [inline layers/cmd_buffer_state.cpp:300]: The lock does need to be held  when object_bindings and broken_bindings are changed.

**Comment by jeremyg-lunarg** [comment]
What is the call stack above CMD_BUFFER_STATE::Destroy()? It seems like that method only holds the write lock for Reset() and then calls BASE_NODE::Destroy() without it held.

**Comment by juan-lunarg** [comment]
> What is the call stack above CMD_BUFFER_STATE::Destroy()? It seems like that method only holds the write lock for Reset() and then calls BASE_NODE::Destroy() without it held.

The issue has the callstack in question:
https://github.com/KhronosGroup/Vulkan-ValidationLayers/issues/5011

**Pull request closed without merging**